### PR TITLE
[For Review] Spark 2.1 PR

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

Just for review, since I can't find a better way.

### CODE REVIEW

**Comment by linhong-intel** [comment]
Reason for "FileFormatWriter extends Serializable"
Caused by: java.io.NotSerializableException: org.apache.spark.sql.execution.datasources.FileFormatWriter
Serialization stack:
- object not serializable (class: org.apache.spark.sql.execution.datasources.FileFormatWriter, value: org.apache.spark.sql.execution.datasources.FileFormatWriter@7856f41a)
- field (class: org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1, name: $outer, type: class org.apache.spark.sql.execution.datasources.FileFormatWriter)
- object (class org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1, <function0>)
- field (class: org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1$$anonfun$3, name: $outer, type: class org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1)
- object (class org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1$$anonfun$3, <function2>)
at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:46)
at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:295)
... 78 more

File: src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+   * Refreshes (or invalidates) any metadata/data cached in the plan recursively.
+   */
+  def refresh(): Unit = children.foreach(_.refresh())
+}
+
+/**
+ * A logical plan node with no children.
+ */
+abstract class LeafNode extends LogicalPlan {
+  override def children: Seq[LogicalPlan] = Nil
```
[the remark this answers is not in the document]
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: I think we already work around this when we upgrade 2.0.0 to 2.0.2. 
Beginning from 2.0.2, we did not override the children, but did this:
override lazy val containsChild: Set[TreeNode[_]] = Seq(relation).toSet
which works as well and no need to change the final property.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 2]
```diff
@@ -48,17 +46,17 @@ case class CreateIndex(
     indexType: AnyIndexType,
     partitionSpec: Option[TablePartitionSpec]) extends RunnableCommand with Logging {
 
-  override lazy val containsChild: Set[TreeNode[_]] = Seq(relation).toSet
+  override def children: Seq[LogicalPlan] = Seq(relation)
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: you can keep this line
override lazy val containsChild: Set[TreeNode[_]] = Seq(relation).toSet.
And if it works well, no need to override the final children of LeafNode(RunnableCommand).

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapReadSupportImpl.scala [hunk 3]
```diff
@@ -50,7 +50,7 @@ import org.apache.spark.sql.execution.datasources.parquet.ParquetReadSupportHelp
  * to pass requested schema from [[init()]]
  * to [[prepareForRead()]], but use a private `var` for simplicity.
  */
-class OapReadSupportImpl extends ReadSupport[InternalRow] with Logging {
+class OapReadSupportImpl extends ReadSupport[UnsafeRow] with Logging {
```
[the remark this answers is not in the document]
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapReadSupportImpl.scala]: I think that is due to the Spark change.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/CacheStatusSerDe.scala [hunk 4]
```diff
@@ -53,12 +53,12 @@ private[oap] object CacheStatusSerDe extends SerDe[String, Seq[FiberCacheStatus]
     compact(render("statusRawDataArray" -> statusJArray))
   }
 
+  private implicit val format = DefaultFormats
+
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/CacheStatusSerDe.scala]: Synced with Songzhan, this is to resolve compile error.

File: src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: 1. In spark 2.1, during resolve relation, before using `containsChild`, `children.nonEmpty` is checked. So `override def children` is necessary
2. This "override" is used to resolve `UnresolvedRelation(tableIdentifier)` to `ResolvedRelation`. But I check spark's similar class, there is an easy way.

Below code is from `case class DescribeTableCommand`
```
val catalog = sparkSession.sessionState.catalog
val relation = catalog.lookupRelation(table)
```

[the remark this answers is not in the document]
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Yes, that's better than change LogicPlan which belongs to SPARK.

File: src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+   * to commit or abort tasks. Exceptions thrown by the implementation of this trait will
+   * automatically trigger task aborts.
+   */
+  trait ExecuteWriteTask {
+    /**
+     * Writes data out to files, and then returns the list of partition strings written out.
+     * The list of partitions is sent back to the driver and used to update the catalog.
+     */
+    def execute(iterator: Iterator[InternalRow]): (Set[String], Seq[WriteResult])
+    def releaseResources(): WriteResult
```
[the remark this answers is not in the document]
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Maybe we can use extra tool to update write result if we don't want to change the interface too much. But that way does not sound good... So far I think this is ok to me.

**Comment by linhong-intel** [comment]
@adrian-wang @yueawang @lee-lei 
Hi,
This PR is ready to merge.
I think we can merge it first. I will keep on reviewing and fixing the code style issues.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/DataSourceMeta.scala [hunk 6]
```diff
@@ -42,12 +42,10 @@ import org.apache.spark.sql.types._
  *    .
  *    .
  * FileMeta N
- * IndexMeta 1      -- 768 bytes
+ * IndexMeta 1      -- 512 bytes
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/DataSourceMeta.scala]: Not forward compatible, but tested in our environment

**Comment by linhong-intel** [comment]
Close this. Use #337 instead

**Pull request closed without merging**